File: python/src/composable_model_graph/math/transfer_function.py
"""Transfer function (LTI / IIR filter), mirroring the TypeScript math package.

A single-input single-output linear time-invariant filter expressed as a rational
transfer function G(q) = B(q)/A(q), i.e. the difference equation (run from rest,
inputs and outputs zero for t < 0):

    y(t) = b0 u(t) + b1 u(t-1) + ... + bnb u(t-nb)
           - a1 y(t-1) - ... - ana y(t-na)

Forward pass only (an IIR filter applied to its input sequence); there is no
backprop here. It is the domain-free *dynamical* counterpart of a dense layer: a
dense layer is a static map, a transfer function carries memory. The coefficients
b, a are exactly what a least-squares ARX fit recovers from input/output data
(linear in the parameters, no autodiff required).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List


@dataclass
class TransferFunction:
    id: str
    name: str
    b: List[float]  # numerator [b0, b1, ..., bnb]; at least b0 is required
    a: List[float] = field(default_factory=list)  # denominator [a1, ..., ana]; a0 = 1 implicit

    def __post_init__(self) -> None:
        if len(self.b) == 0:
            raise ValueError(
                "TransferFunction requires at least one numerator coefficient (b0)"
            )

    def run(self, u: List[float]) -> List[float]:
        n = len(u)
        y = [0.0] * n
        for t in range(n):
            acc = 0.0
            for j, bj in enumerate(self.b):
                k = t - j
                if k >= 0:
                    acc += bj * u[k]
            for i, ai in enumerate(self.a):
                k = t - (i + 1)
                if k >= 0:
                    acc -= ai * y[k]
            y[t] = acc
        return y


def create_transfer_function(
    id: str, name: str, b: List[float], a: List[float] | None = None
) -> TransferFunction:
    return TransferFunction(id=id, name=name, b=list(b), a=list(a or []))

File: typescript/packages/math/src/transfer-function.ts
import type { Transform } from "@composable-model-graph/core";

/** Configuration for a {@link TransferFunction}. */
export interface TransferFunctionConfig {
  id: string;
  name: string;
  /** Numerator coefficients [b0, b1, ..., bnb] of B(q). At least b0 is required. */
  b: number[];
  /** Denominator coefficients [a1, ..., ana] of A(q); a0 = 1 is implicit. Empty = FIR. */
  a: number[];
}

/**
 * A single-input single-output linear time-invariant (LTI) filter expressed as a
 * rational transfer function and run as a core {@link Transform}.
 *
 *   G(q) = B(q) / A(q),   A(q) = 1 + a1 q^-1 + ...,   B(q) = b0 + b1 q^-1 + ...
 *
 * which is the linear difference equation (the filter starts from rest, i.e. inputs
 * and outputs are zero for t < 0):
 *
 *   y(t) = b0 u(t) + b1 u(t-1) + ... + bnb u(t-nb)
 *          - a1 y(t-1) - ... - ana y(t-na)
 *
 * This is the forward pass only (an IIR filter applied to its input sequence); there
 * is no backprop here. It is the domain-free *dynamical* counterpart of
 * {@link DenseLayer}: DenseLayer is a static map, TransferFunction carries memory.
 * The coefficients b, a are exactly what a least-squares ARX fit recovers from
 * input/output data (linear in the parameters, no autodiff required).
 */
export class TransferFunction implements Transform<number[], number[]> {
  readonly id: string;
  readonly name: string;
  readonly b: number[];
  readonly a: number[];

  constructor(config: TransferFunctionConfig) {
    if (config.b.length === 0) {
      throw new Error(
        "TransferFunction requires at least one numerator coefficient (b0)",
      );
    }
    this.id = config.id;
    this.name = config.name;
    this.b = config.b;
    this.a = config.a;
  }

  run(input: number[]): number[] {
    const T = input.length;
    const y = new Array<number>(T).fill(0);
    for (let t = 0; t < T; t++) {
      let acc = 0;
      for (let j = 0; j < this.b.length; j++) {
        const k = t - j;
        if (k >= 0) acc += this.b[j]! * input[k]!;
      }
      for (let i = 0; i < this.a.length; i++) {
        const k = t - (i + 1);
        if (k >= 0) acc -= this.a[i]! * y[k]!;
      }
      y[t] = acc;
    }
    return y;
  }
}

/** Convenience factory mirroring the other `create*` helpers. */
export function createTransferFunction(config: TransferFunctionConfig): TransferFunction {
  return new TransferFunction(config);
}
